### PR TITLE
Add option to disable ExternalIP mitigation (CVE-2020-8554).

### LINKED ISSUE #28187: Regression in DSR for some setups
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We're using barebone dedicated servers on hetzner for our kubernetes cluster and out setup is:
* each server has static ip (L3 connectivity) - wan address, main ip for the server
* additionaly each server is connected to L2 internal network (called vswitch)
* several wan IP's are assigned to this L2 network - they are assigned/reassigned to servers by keepalived and used as 'load balancer' (i.e. they are used as external-ips to connect to ingress)

DSR is partially broken in some scenarios after following commit: https://github.com/cilium/cilium/commit/78f93188ae0e4a74c3794b517047e33b9c3984c5

Connectivity is broken in the following case: pod  1on node 1 connects to some service via external dns (which binds to external ips) and connect choose external-ip which is assigned to node 2 and service load balancer choose ingress pod (let it be pod 2) which is located exactlt on node 1. In this case pod 2 answer will have destination ip address which is locally assigned to node 1. It was all OK before moving DSR logic out of lxc - source ip address was changed to dsr one and packet was processed. After the change source ip address is unchanged after lxc and packet is consumed by localhost stack - and connectivity is broken.

* pod1 sends packet, src = pod1 internall address on node1, dst = external-ip on node2
* node1 performs masquerading to send packet to wan, src = node1 wan address, dst = external-ip on node2
* node2 receives packet and performs load balancing to service, src = node1 wan address, dst = pod2 internall address on node1
* node1 received packet and it is consumed by pod2, src = node 1 wan address, dst = pod2 internall address on node1
* pod2 sends answer, src = pod2 internall address on node1, dst = node 1 wan address

before change it was:
* node1 performs dsr, src = external-ip on node2, dst = node 1 wan address
* node1 receive packet and performs unmasquerading, src = external-ip on node2, dst = pod 1 internall address on node1
* pod1 receive answer

after change:
* after lxc processing there is no dsr and packet is treated as local and it never gets to to-netdev -> connectivity is broken

### Cilium Version

>= 1.13.3

### Kernel Version

Any, not related.
Tested on 6.5.1.

### Kubernetes Version

Any, not related.
Tested on 1.27.4.

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by kvaster**
This mitigation has it's own drawbacks for some setups. It prevents pods communication in same cluster via ExternalIP when DSR is enabled.

Fixes #28187

### CODE REVIEW

**Review by lmb — APPROVED** [review]

**Review by gandro — APPROVED** [review]
Helm/Agent looks good to me!

**Review by borkmann — APPROVED** [review]
lgtm

**Review by derailed — APPROVED** [review]
@kvaster LGTM - Thank you for this update!

File: bpf/bpf_sock.c [hunk 1]
```diff
@@ -156,10 +156,12 @@ static __always_inline int sock4_update_revnat(struct bpf_sock_addr *ctx,
...
 static __always_inline bool
-sock4_skip_xlate(struct lb4_service *svc, __be32 address)
+sock4_skip_xlate(struct lb4_service *svc, __be32 address __maybe_unused)
 {
 	if (lb4_to_lb6_service(svc))
 		return true;
+
+#ifndef DISABLE_EXTERNAL_IP_MITIGATION
 	if (lb4_svc_is_external_ip(svc) ||
 	    (lb4_svc_is_hostport(svc) && !is_v4_loopback(address))) {
```
Comment (julianwiedmann) [inline bpf/bpf_sock.c]: We only want to target the `lb4_svc_is_external_ip(svc)` check though, right? The `lb4_svc_is_hostport(svc)` check should not be in scope?

Also - does the IPv6 path need the same change?

File: daemon/cmd/daemon_main.go [hunk 2]
```diff
@@ -1016,6 +1016,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.StringSlice(option.VLANBPFBypass, []string{}, "List of explicitly allowed VLAN IDs, '0' id will allow all VLAN IDs")
 	option.BindEnv(vp, option.VLANBPFBypass)
 
+	flags.Bool(option.DisableExternalIPMitigation, false, "Disable ExternalIP mitigation (CVE-2020-8554, default false)")
+	option.BindEnv(vp, option.DisableExternalIPMitigation)
+
```
Comment (julianwiedmann) [inline daemon/cmd/daemon_main.go:1020]: Could you mention this option in https://github.com/cilium/cilium/blob/main/Documentation/network/kubernetes/kubeproxy-free.rst - the DSR section maybe?

**Review by julianwiedmann — CHANGES_REQUESTED** [review]
Thank you for the fix! Just two smaller aspects worth addressing.

**Review by joamaki — APPROVED** [review]

File: bpf/bpf_sock.c (hunk 1 above)
Comment (kvaster) [reply bpf/bpf_sock.c]: Telling the the truth, I was not aware of ipv6, cause we're ipv4 only currently. But I will check this and will get back.

File: daemon/cmd/daemon_main.go (hunk 2 above)
Comment (kvaster) [reply daemon/cmd/daemon_main.go:1020]: Will make change and will get back.

File: bpf/bpf_sock.c (hunk 1 above)
Comment (kvaster) [reply bpf/bpf_sock.c]: @julianwiedmann , you're right for both notes. We should only target for `lb4_svc_is_external_ip(svc)`. And we should apply the same for IPv6 aswell. Fixed.

File: daemon/cmd/daemon_main.go (hunk 2 above)
Comment (kvaster) [reply daemon/cmd/daemon_main.go:1020]: Fixed.

**Review by danehans — APPROVED** [review]

File: bpf/bpf_sock.c [hunk 3]
```diff
@@ -160,7 +160,11 @@ sock4_skip_xlate(struct lb4_service *svc, __be32 address)
 {
 	if (lb4_to_lb6_service(svc))
 		return true;
-	if (lb4_svc_is_external_ip(svc) ||
+
+	if (
+#ifndef DISABLE_EXTERNAL_IP_MITIGATION
+		lb4_svc_is_external_ip(svc) ||
+#endif
```
Comment (julianwiedmann) [inline bpf/bpf_sock.c]: As the BPF stylechecker is complaining about this pattern, one suggestion to make it a bit cleaner:

```suggestion
	if ((lb4_svc_is_external_ip(svc) && !is_defined(DISABLE_EXTERNAL_IP_MITIGATION))) ||
```

**Review by julianwiedmann — COMMENTED** [review]
lgtm so far, ty! One suggestion below to get the BPF stylechecker :heavy_check_mark: .

File: Documentation/network/kubernetes/kubeproxy-free.rst [hunk 4]
```diff
@@ -490,6 +490,10 @@ In some public cloud provider environments that implement source /
 destination IP address checking (e.g. AWS), the checking has to be disabled in
 order for the DSR mode to work.
 
+By default cilium uses special ExternalIP mitigation CEV-2020-8554 MITM vulnerability.
```
Comment (ghost) [inline Documentation/network/kubernetes/kubeproxy-free.rst]:
```suggestion
By default Cilium uses special ExternalIP mitigation for CVE-2020-8554 MITM vulnerability.
```

**Review by ghost — APPROVED** [review]
Left one nit comment for docs, but looks good otherwise.

File: bpf/bpf_sock.c (hunk 3 above)
Comment (kvaster) [reply bpf/bpf_sock.c]: Fixed

File: Documentation/network/kubernetes/kubeproxy-free.rst (hunk 4 above)
Comment (kvaster) [reply Documentation/network/kubernetes/kubeproxy-free.rst]: Fixed

**Review by ysksuzuki — APPROVED** [review]

**Comment by kvaster** [comment]
Telling the truth, I don't know why helm charts test is failing... I've rechecked it twice...

**Comment by julianwiedmann** [comment]
> Telling the truth, I don't know why helm charts test is failing... I've rechecked it twice...

https://github.com/cilium/cilium/actions/runs/8611642349/job/23599228700?pr=31513#step:3:111 is what's needed:
>please run 'make -C install/kubernetes' and submit your changes

(it updates `install/kubernetes/cilium/values.schema.json`)

**Review by julianwiedmann — APPROVED** [review]
lgtm, thank you!

**Pull request merged**